filebeat/processor/add_kubernetes_metadata/matchers.go

```diff
@@ -78,59 +78,79 @@ func newLogsPathMatcher(cfg common.Config) (add_kubernetes_metadata.Matcher, err
+		return ""
+	}
 
-		sourceLen := len(source)
-		logsPathLen := len(f.LogsPath)
+	sourceLen := len(source)
+	logsPathLen := len(f.LogsPath)
 
-		if f.ResourceType == "pod" {
-			// Specify a pod resource type when manually mounting log volumes and they end up under "/var/lib/kubelet/pods/"
-			// This will extract only the pod UID, which offers less granularity of metadata when compared to the container ID
-			if strings.HasPrefix(f.LogsPath, podLogsPath()) && strings.HasSuffix(source, ".log") {
+	if f.ResourceType == "pod" {
+		// Pod resource type will extract only the pod UID, which offers less granularity of metadata when compared to the container ID
+		if strings.HasSuffix(source, ".log") {
+			// Specify a pod resource type when writting logs into manually mounted log volume,
+			// those logs apper under under "/var/lib/kubelet/pods/<pod_id>/volumes/..."
+			if strings.HasPrefix(f.LogsPath, podKubeletLogsPath()) {
 				pathDirs := strings.Split(source, pathSeparator)
+				podUIDPos := 5
 				if len(pathDirs) > podUIDPos {
 					podUID := strings.Split(source, pathSeparator)[podUIDPos]
-
 					f.logger.Debugf("Using pod uid: %s", podUID)
 					return podUID
 				}
-
-				f.logger.Error("Error extracting pod uid - source value contains matcher's logs_path, however it is too short to contain a Pod UID.")
-			}
-		} else {
-			// In case of the Kubernetes log path "/var/log/containers/",
-			// the container ID will be located right before the ".log" extension.
-			if strings.HasPrefix(f.LogsPath, containerLogsPath()) && strings.HasSuffix(source, ".log") && sourceLen >= containerIdLen+4 {
-				containerIDEnd := sourceLen - 4
-				cid := source[containerIDEnd-containerIdLen : containerIDEnd]
-				f.logger.Debugf("Using container id: %s", cid)
-				return cid
 			}
-
-			// In any other case, we assume the container ID will follow right after the log path.
-			// However we need to check the length to prevent "slice bound out of range" runtime errors.
-			if sourceLen >= logsPathLen+containerIdLen {
-				cid := source[logsPathLen : logsPathLen+containerIdLen]
-				f.logger.Debugf("Using container id: %s", cid)
-				return cid
+			// In case of the Kubernetes log path "/var/log/pods/",
+			// the pod ID will be extracted from the directory name,
+			// file name example: "/var/log/pods/'<namespace>_<pod_name>_<pod_uid>'/container_name/0.log".
+			if strings.HasPrefix(f.LogsPath, podLogsPath()) {
+				pathDirs := strings.Split(source, pathSeparator)
+				podUIDPos := 4
+				if len(pathDirs) > podUIDPos {
+					podUID := strings.Split(pathDirs[podUIDPos], "_")
+					if len(podUID) > 2 {
+						f.logger.Debugf("Using pod uid: %s", podUID)
+						return podUID[2]
+					}
+				}
 			}
 
-			f.logger.Error("Error extracting container id - source value contains matcher's logs_path, however it is too short to contain a Docker container ID.")
+			f.logger.Error(`Error extracting pod uid - source value does not contains matcher's logs_path,
```

Comment:
could you please elaborate on this? do you mean do smth similar to other processors. like - https://github.com/elastic/beats/blob/master/libbeat/processors/add_kubernetes_metadata/config.go ?

Reply:
Yes, in https://github.com/elastic/beats/blob/3d19ec04aaf73b20eafb9d5283a43c17543d8b4f/libbeat/processors/add_kubernetes_metadata/config.go#L63 we can check if `matchers`' config is valid too so as to warn the users early on and avoid initialising the processor in case the config is not supported. I see that the matcher's config part is a map so not sure how much mangling this change would require but it should be doable. Let me know what you think.

Reply:
I will give it a try

Reply:
@ChrsMark please have a look - https://github.com/elastic/beats/pull/28868/commits/ab3c82161a433583dabb68e3499aa83e22990e21, let me know wdyt, are there any better ways maybe?